First basic deployment

nicolaferraro opened the pull request:
This does a first deployment of upstream Syndesis (Fixes #2).

After building it, install the operator, then put a route name in `deploy/cr.yaml` and create the resource. All Syndesis objects will be created. The template is not created, it's just processed on the fly.

From the scaffolding, I've added the `tmp` dir back into source version because it contains scripts needed by operator sdk. In general, I've used the `.gitignore` provided by operator-sdk with few changes.

I'm using my fork of the operator-sdk because I need a small change to get the rest configuration in order to instantiate the Openshift template processor. I'll try to merge it back into the sdk repo.

The `template.yaml` and the other file have been changed (temporarily) because I had to set the correct group for legacy `/oapi` resources (e.g. `v1` --> `apps.openshift.io/v1` for `DeploymentConfig`). I think we can fix it in the operator, there should be a way to do it in Openshift libraries.

I've used `go-bindata` to embed resources in the final program. This requires that you regenerate the `bindata.go` file in the `resources` dir every time you update the yaml files.

Some issues left:
- What if the user installs 2 Syndesis resources in the same namespace? We should create only one?
- Configuration mapping from CR fields to TEMPLATE_PARAMETERS has not been completed
- Detection of whether Syndesis is running is a bit basic
- Version selection (should we bundle different templates?)
- State management (maybe we can display some state in the shell for `oc get syndesis`)
- Tests... as soon as I load some sdk package, the init scripts require that I'm connected to a Kubernetes instance... Should we setup tests with Openshift?

nicolaferraro committed:
Adding back boilerplate code needed by operator-sdk

nicolaferraro committed:
First working version with basic new deployment

rhuss commented:
Thanks ! will have a look ASAP.

> What if the user installs 2 Syndesis resources in the same namespace? We should create only one?

yes, only one instance of Syndesis per namespace. There should be an error reported if a second resoure is created. Not sure how this can be dealt with on the K8s level. E.g. whether the operator has to delete the resource if it detects a second one. Or maybe better, we set it in some specific error status which shows up in 'oc get' or 'oc describe'. Not sure how "singleton" are handled anyway. Maybe a good question to the coreos community ? What are the best channels to ask for that kind of questions btw ?

> Configuration mapping from CR fields to TEMPLATE_PARAMETERS has not been completed

I can take this over.

> Version selection (should we bundle different templates?)

 I think we are fine, if we  include only a single template. Each operator should match the version of the app it installs. At least that's what has been suggested to me by @robszumski

> State management (maybe we can display some state in the shell for oc get syndesis)

Yes, we need to think about the `status:` section and what kind of status we want to support.

> Tests... as soon as I load some sdk package, the init scripts require that I'm connected to a Kubernetes instance... Should we setup tests with Openshift?

Not sure how to unit test operator right now. @robszumski do you have any pointers how to test operators with unit and integration tests ?

nicolaferraro committed:
Remove go-bindata from dependencies

nicolaferraro committed:
Workaround for dep bug

nicolaferraro committed:
Removed dependencies on kubernetes and openshift

nicolaferraro committed:
Refactored control logic into a statemachine

nicolaferraro commented:
@rhuss I've removed dependencies on Openshift/Kubernetes, used a direct REST client.
The "dynamic client" cannot be used, since features that we need for processing templates are available in the "dynamic client" for Kubernetes 1.11, and we can't upgrade since the operator sdk depends on code from client 1.9.

I've added a statemachine that manages the following states:
- None
- Installing
- Starting
- Installed
- NotInstalled (in case of multiple resources on the same namespace, all but one get this state)

There are additional states that we may want to handle later, especially the ones related to upgrades.

nicolaferraro committed:
Added management of serviceaccount secret for authentication

nicolaferraro commented:
Now the operator manages the service account correctly..

nicolaferraro committed:
Completed environment variables configuration from CRD

nicolaferraro committed:
Added some hacks for code generation

rhuss commented on pkg/syndesis/reconcile.go:
```diff
@@ -1,10 +1,37 @@
+func init() {
+	actionPool = append(actionPool,
+		&action.Initialize{},
+		&action.Install{},
```
I wonder whether we  need to distinguish those two phases or whether we shouldn't combine initialize and install ? Before doing the install we could do the sanity check that only a single `Syndesis` is installed.

nicolaferraro replied:
So, I've added the `initialize` stage in order to have a initial step to execute as soon as the CR is created. I was thinking that the next state may be "Installing" or "Upgrading" or "Reconciling", depending on what we'll find out when we'll work on "upgrade from operator" or "upgrade from a previous template installation". We may remove it if it will not be useful.

Putting a state diagram in the readme may be useful..

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+
+	logrus.Info("Installing Syndesis resource ", syndesis.Name)
+
+	serviceAccountRes, err := util.LoadKubernetesResourceFromAsset("oauth-client-sa.yaml")
```
We might want to create the SA directly with the client without reading in a file first. Would make it a bit simpler, I guess.

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+	if sa, ok := serviceAccountRes.(*coreV1.ServiceAccount); ok {
+		saName = sa.Name
+	} else {
+		return errors.New("Cannot determine service account name")
```
acccording to go convention the error message must not start with a capital letter.

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+		return errors.New("Cannot determine service account name")
+	}
+
+	customizeKubernetesResource(serviceAccountRes, syndesis)
```
maybe better name `setNamespaceAndOwnerReference` ? (just to make it a bit more specific)

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+	params["ROUTE_HOSTNAME"] = syndesis.Spec.RouteHostName
+	params["OPENSHIFT_PROJECT"] = syndesis.Namespace
+	params["OPENSHIFT_OAUTH_CLIENT_SECRET"] = token
+
```
here we should copy over the remaining params from `syndesis` (you mentioned that already, just a reminder). Probably best in a dedicated method.

nicolaferraro replied:
Yes, I've already started working on the refactoring :smile:

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+		customizeKubernetesResource(res, syndesis)
+
+		err = sdk.Create(res)
+		if err != nil && !k8serrors.IsAlreadyExists(err) {
```
Not sure whether we should continue when the object already exists. Shouldn't we better delete it and recreate it ?

nicolaferraro replied:
Probably yes. If a installation fails in the middle, by reprocessing the template we get different values for generated parameters, so we may have out of sync resources. Recreating them seems a better option.
But I'm not sure if we should do it for all resources, e.g. PVC...

rhuss commented on pkg/syndesis/action/install.go:
```diff
@@ -0,0 +1,96 @@
+	}
+
+	// Installation completed, set the next state
+	target := syndesis.DeepCopy()
```
A question because I'm a go noop. Why are we doing here a `DeepCopy` ? Is it because the event could be reused for multiple handlers ? Wouldn't it be better to already reflect its new state when handed over to the other handlers ? Not sure what the best practices are here ...

nicolaferraro replied:
Don't know if there's a convention. But from a resource management point of view, since the actions need to change the status only few times during the life of the CR, but the callback is invoked by the sdk every `x` seconds, I preferred to do deepcopy only when it needs to be changed.

rhuss commented on pkg/syndesis/action/startup.go:
```diff
@@ -0,0 +1,58 @@
+	return syndesisInstallationStatusIs(syndesis, v1alpha1.SyndesisInstallationStatusStarting)
+}
+
+func (a *Startup) Execute(syndesis *v1alpha1.Syndesis) error {
```
Isn't this event handler only called once when the status changes from 'Installing' to 'Starting'. When the DCs are not ready then (very likely), when would be this method called be next ? Only when something other changes in the `Syndesis` resource, so I suspect that the status `Installed` is not easily reached. 

I think, instead of listening on `Syndesis` resources, we should listen on `DeploymentConfigs` directly. When such an even comes in, the handler should do the query on the *other* dcs and then when everything is started, fetch the Syndesis resource and update the status from there.

This would be also working when a DC drops out, so the syndesis status could be set to `not running`.

BTW, I would use something liken `Running` or `CrashLoopBack` (or so, sth similar of the DC status) here  to indidicate the runtime status for the application.

Said that, we could also keep a list of all components with their status directly in the Syndesis status, so one can easily see which component would be in an error state without doing a query over all DCs (just a nice add on I guess)

nicolaferraro replied:
I think listening to the DCs is not necessary. The SDK is using a `resyncPeriod` of 5 seconds, so that even if we miss a change event, we are asked later to resync. For example, the first time I create the service account, tokens are usually not available immediately but I need to wait 1-2 resync periods before they are available and I can use them in the template (I was thinking to split the "Installing" phase in two subphases to have a better handling, btw it currently works).

So, unless we want immediate notification when Syndesis is ready, we can wait for the resync trigger.

Tracking the status can be done on the Syndesis status too. It may also be another action that is executed no matter what is the current "InstallationStatus".

rhuss replied:
So the handler is called for every re-sync, so every 5 seconds, even when there has been no change ? If so, I'm fine with that, but then we should include a state like 'CrashLoopback' or sth. when going from 'Running' to broken, i.e. when there are less than expected DCs scaled to 1.

rhuss reviewed (commented):
Awesome, looks good !

I did a first review on the top-level methods, and have some comments and questions, but it mostly boils down to:

* Wouldn't it be better to listen to DC events for determing when syndesis is running ? (see comment below)
* What's about the overall error handling ? What happens when `Reconcile` returns with an error ?

nicolaferraro committed:
Remove space from omitempty tag

robszumski commented:
> What are the best channels to ask for that kind of questions btw ?

Join us on the mailing list: https://groups.google.com/forum/#!forum/operator-framework

> Not sure how to unit test operator right now. @robszumski do you have any pointers how to test operators with unit and integration tests ?

We have plans to add functionality into the SDK to make this easier to execute. More details: https://github.com/operator-framework/operator-sdk/blob/master/doc/design/long-term/testing-operators.md

nicolaferraro committed:
Refactoring after review

nicolaferraro committed:
Restrict startup to infrastructure deployments

nicolaferraro committed:
Reporting startup failures in Syndesis resource

nicolaferraro committed:
Renaming script dir to hack

nicolaferraro committed:
Updating dependencies
